File: android/core/src/main/kotlin/com/danielhipskind/fihaven/core/net/ApiClient.kt
package com.danielhipskind.fihaven.core.net

import com.danielhipskind.fihaven.core.model.AppData
import com.danielhipskind.fihaven.core.model.FiHavenJson
import com.danielhipskind.fihaven.core.model.Entitlement
import com.danielhipskind.fihaven.core.model.PromoResult

/// Talks to the FiHaven REST API with token/Bearer auth
/// (docs/native-contract.md §3–5). Mirrors the Swift APIClient.
class ApiClient(
    private val config: ApiConfig,
    private val tokens: TokenStore,
    private val transport: HttpTransport = DefaultHttpTransport(),
) {
    // ── request building (internal so tests can assert it) ───────────
    internal fun makeRequest(
        path: String,
        method: HttpMethod,
        body: String? = null,
        tokenMode: Boolean = false,
    ): HttpRequest {
        val headers = mutableMapOf<String, String>()
        tokens.get()?.let { headers["Authorization"] = "Bearer $it" }
        if (tokenMode) headers["X-Auth-Mode"] = "token"
        if (body != null) headers["Content-Type"] = "application/json"
        return HttpRequest(method, config.baseUrl.trimEnd('/') + "/" + path, headers, body)
    }

    private suspend fun send(request: HttpRequest): String {
        val resp = try {
            transport.send(request)
        } catch (e: ApiError) {
            throw e
        } catch (e: Exception) {
            throw ApiError.Transport(e.message ?: "transport error")
        }
        if (resp.status == 401) throw ApiError.Unauthenticated
        if (resp.status !in 200..299) {
            val code = runCatching {
                FiHavenJson.decodeFromString(ErrorBody.serializer(), resp.body).error
            }.getOrNull()
            throw ApiError.Http(resp.status, code)
        }
        return resp.body
    }

    private inline fun <reified T> decode(body: String): T =
        try {
            FiHavenJson.decodeFromString<T>(body)
        } catch (e: Exception) {
            throw ApiError.Decoding(e.message ?: "decode error")
        }

    private inline fun <reified T> encode(value: T): String = FiHavenJson.encodeToString(value)

    // ── Auth ─────────────────────────────────────────────────────────
    suspend fun signup(email: String, password: String, captchaToken: String, loginStartedAt: Long): AuthSession {
        val body = encode(LoginRequest(email, password, captchaToken, loginStartedAt))
        return storeSession(send(makeRequest("api/auth/signup", HttpMethod.POST, body, tokenMode = true)))
    }

    suspend fun login(email: String, password: String, captchaToken: String, loginStartedAt: Long): LoginOutcome {
        val body = encode(LoginRequest(email, password, captchaToken, loginStartedAt))
        val response = send(makeRequest("api/auth/login", HttpMethod.POST, body, tokenMode = true))
        val mfa = runCatching { decode<MfaResponse>(response) }.getOrNull()
        if (mfa?.mfaRequired == true) {
            return LoginOutcome.MfaRequired(MfaChallenge(mfa.mfaToken ?: "", mfa.methods ?: emptyList()))
        }
        return LoginOutcome.Authenticated(storeSession(response))
    }

    suspend fun verifyMfa(mfaToken: String, code: String): AuthSession {
        val body = encode(MfaVerifyRequest(mfaToken, code))
        return storeSession(send(makeRequest("api/auth/mfa/verify", HttpMethod.POST, body, tokenMode = true)))
    }

    suspend fun sendEmailCode(mfaToken: String) {
        send(makeRequest("api/auth/mfa/email/send", HttpMethod.POST, encode(MfaTokenRequest(mfaToken))))
    }

    suspend fun me(): User? = decode<MeResponse>(send(makeRequest("api/auth/me", HttpMethod.GET))).user

    /** Re-send the email-verification message for the current session. */
    suspend fun resendVerification() {
        send(makeRequest("api/auth/resend-verification", HttpMethod.POST))
    }

    /** Mark first-run onboarding complete (CSRF is skipped for Bearer auth). */
    suspend fun markOnboarded() {
        send(makeRequest("api/account/onboarded", HttpMethod.POST))
    }

    // ── Plaid (bank linking, Pro-gated) ──────────────────────────────
    suspend fun plaidStatus(): PlaidStatus =
        decode(send(makeRequest("api/plaid/status", HttpMethod.GET)))

    // Pass itemId for an update-mode token (re-auth an existing item).
    suspend fun plaidLinkToken(itemId: Int? = null): String {
        val req = if (itemId != null)
            makeRequest("api/plaid/link/token", HttpMethod.POST, encode(PlaidLinkTokenBody(itemId)))
        else makeRequest("api/plaid/link/token", HttpMethod.POST)
        return decode<PlaidLinkTokenResponse>(send(req)).linkToken
    }

    suspend fun plaidExchange(publicToken: String) {
        send(makeRequest("api/plaid/link/exchange", HttpMethod.POST, encode(PlaidExchangeBody(publicToken))))
    }

    // After a successful update-mode Link, mark the item repaired.
    suspend fun plaidRepaired(itemId: Int) {
        send(makeRequest("api/plaid/item/$itemId/repaired", HttpMethod.POST))
    }

    suspend fun plaidRemove(itemId: Int) {
        send(makeRequest("api/plaid/item/$itemId/remove", HttpMethod.POST))
    }

    suspend fun plaidRefresh(): List<PlaidItem> =
        decode<PlaidItemsResponse>(send(makeRequest("api/plaid/refresh", HttpMethod.POST))).items

    suspend fun logout() {
        runCatching { send(makeRequest("api/auth/logout", HttpMethod.POST)) }
        tokens.clear()
    }

    private fun storeSession(body: String): AuthSession {
        val r = decode<SessionResponse>(body)
        val token = r.token ?: throw ApiError.Decoding("missing token in auth response")
        tokens.set(token)
        return AuthSession(token, r.user)
    }

    // ── Data sync ─────────────────────────────────────────────────────
    suspend fun fetchData(): AppData = decode(send(makeRequest("api/data", HttpMethod.GET)))

    suspend fun saveData(data: AppData) {
        val body = encode(DataPutBody(data.bills, data.cards, data.payments, data.settings))
        send(makeRequest("api/data", HttpMethod.PUT, body))
    }

    // ── Billing / entitlement ─────────────────────────────────────────
    suspend fun billingStatus(): Entitlement =
        decode<EntitlementResponse>(send(makeRequest("api/billing/status", HttpMethod.GET))).entitlement

    suspend fun verifyGoogle(productId: String, purchaseToken: String, expiryTimeMillis: Long? = null): Entitlement =
        decode<EntitlementResponse>(
            send(makeRequest("api/billing/google/verify", HttpMethod.POST,
                encode(GoogleVerifyBody(productId, purchaseToken, expiryTimeMillis))))
        ).entitlement

    /** Redeem a server promo code. Throws ApiError.Http(409, …) for
     *  invalid / exhausted / already-redeemed. */
    suspend fun redeemPromo(code: String): PromoResult =
        decode(send(makeRequest("api/billing/promo/redeem", HttpMethod.POST, encode(PromoRedeemBody(code)))))

    // ── Profile ──────────────────────────────────────────────────────
    suspend fun changeName(name: String): String? =
        decode<NameResult>(send(makeRequest("api/account/change-name", HttpMethod.POST, encode(ChangeNameBody(name))))).name

    suspend fun changeEmail(password: String, newEmail: String): String? =
        decode<EmailResult>(send(makeRequest("api/account/change-email", HttpMethod.POST, encode(ChangeEmailBody(password, newEmail))))).email

    suspend fun changePassword(currentPassword: String, newPassword: String) {
        send(makeRequest("api/account/change-password", HttpMethod.POST, encode(ChangePasswordBody(currentPassword, newPassword))))
    }

    suspend fun deleteAccount(password: String) {
        send(makeRequest("api/account/delete", HttpMethod.POST, encode(PasswordBody(password))))
    }

    suspend fun exportData(): String = send(makeRequest("api/account/export", HttpMethod.GET))

    // ── MFA ───────────────────────────────────────────────────────────
    suspend fun mfaStatus(): MfaStatus = decode(send(makeRequest("api/account/mfa/status", HttpMethod.GET)))

    suspend fun totpSetup(password: String): TotpSetup =
        decode(send(makeRequest("api/account/mfa/totp/setup", HttpMethod.POST, encode(PasswordBody(password)))))

    suspend fun totpConfirm(code: String): List<String> =
        decode<BackupCodesResult>(send(makeRequest("api/account/mfa/totp/confirm", HttpMethod.POST, encode(CodeBody(code))))).backupCodes

    suspend fun totpDisable(password: String, code: String) {
        send(makeRequest("api/account/mfa/totp/disable", HttpMethod.POST, encode(PasswordCodeBody(password, code))))
    }

    suspend fun regenerateBackupCodes(password: String, code: String): List<String> =
        decode<BackupCodesResult>(send(makeRequest("api/account/mfa/backup-codes/regenerate", HttpMethod.POST, encode(PasswordCodeBody(password, code))))).backupCodes

    suspend fun emailMfaEnable(password: String): String =
        decode<EmailEnableResult>(send(makeRequest("api/account/mfa/email/enable", HttpMethod.POST, encode(PasswordBody(password))))).challengeId

    suspend fun emailMfaConfirm(challengeId: String, code: String) {
        send(makeRequest("api/account/mfa/email/confirm", HttpMethod.POST, encode(EmailConfirmBody(challengeId, code))))
    }

    suspend fun emailMfaDisable(password: String) {
        send(makeRequest("api/account/mfa/email/disable", HttpMethod.POST, encode(PasswordBody(password))))
    }

    suspend fun listPasskeys(): List<PasskeyInfo> =
        decode<PasskeyListResult>(send(makeRequest("api/account/mfa/passkey/list", HttpMethod.GET))).passkeys

    suspend fun deletePasskey(id: Int, password: String) {
        send(makeRequest("api/account/mfa/passkey/delete", HttpMethod.POST, encode(PasskeyDeleteBody(id, password))))
    }

    companion object {
        fun now(): Long = System.currentTimeMillis()
    }
}

File: ios/FiHavenCore/Sources/FiHavenCore/Networking/APIClient.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking  // URLSession lives here on Linux
#endif

/// Type-erased Encodable so `makeRequest` can take any body.
struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void
    init(_ wrapped: Encodable) { _encode = { try wrapped.encode(to: $0) } }
    func encode(to encoder: Encoder) throws { try _encode(encoder) }
}

/// Talks to the FiHaven REST API using token/Bearer auth
/// (docs/native-contract.md §3–4). Stateless apart from the injected
/// `TokenStore`; safe to share.
public final class APIClient: Sendable {
    public enum Method: String {
        case GET, POST, PUT, DELETE
    }

    // Module-internal so endpoint extensions (e.g. APIClient+Household) can
    // build their own requests (the SSE stream needs the raw session).
    let config: APIConfig
    let tokens: TokenStore
    let session: URLSession

    public init(config: APIConfig, tokens: TokenStore, session: URLSession = .shared) {
        self.config = config
        self.tokens = tokens
        self.session = session
    }

    // ── Request building (internal so tests can assert it) ───────────

    func makeRequest(
        path: String,
        method: Method,
        body: AnyEncodable? = nil,
        tokenMode: Bool = false
    ) throws -> URLRequest {
        let url = config.baseURL.appendingPathComponent(path)
        var req = URLRequest(url: url)
        req.httpMethod = method.rawValue
        if let token = tokens.get() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if tokenMode {
            req.setValue("token", forHTTPHeaderField: "X-Auth-Mode")
        }
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(body)
        }
        return req
    }

    /// Sends a request, mapping non-2xx into APIError. 401 → `.unauthenticated`.
    /// `internal` so the account/MFA extension can reuse it.
    @discardableResult
    func send(_ req: URLRequest) async throws -> Data {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw APIError.transport(error.localizedDescription)
        }
        guard let http = response as? HTTPURLResponse else {
            throw APIError.transport("Non-HTTP response")
        }
        if http.statusCode == 401 { throw APIError.unauthenticated }
        guard (200..<300).contains(http.statusCode) else {
            let code = (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
            throw APIError.http(status: http.statusCode, code: code)
        }
        return data
    }

    func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw APIError.decoding(String(describing: error))
        }
    }

    // ── Auth ─────────────────────────────────────────────────────────

    /// Epoch-ms timestamp for the anti-bot `loginStartedAt` field. Capture
    /// when the auth screen appears; submit must be ≥ 2500 ms later.
    public static func now() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }

    public func signup(
        email: String,
        password: String,
        captchaToken: String,
        loginStartedAt: Int64
    ) async throws -> AuthSession {
        let body = LoginRequest(
            email: email, password: password,
            captchaToken: captchaToken, loginStartedAt: loginStartedAt, website: ""
        )
        let req = try makeRequest(path: "api/auth/signup", method: .POST,
                                  body: AnyEncodable(body), tokenMode: true)
        let data = try await send(req)
        return try storeSession(from: data)
    }

    public func login(
        email: String,
        password: String,
        captchaToken: String,
        loginStartedAt: Int64
    ) async throws -> LoginOutcome {
        let body = LoginRequest(
            email: email, password: password,
            captchaToken: captchaToken, loginStartedAt: loginStartedAt, website: ""
        )
        let req = try makeRequest(path: "api/auth/login", method: .POST,
                                  body: AnyEncodable(body), tokenMode: true)
        let data = try await send(req)

        if let mfa = try? JSONDecoder().decode(MfaResponse.self, from: data),
           mfa.mfaRequired == true {
            return .mfaRequired(MfaChallenge(
                mfaToken: mfa.mfaToken ?? "",
                methods: mfa.methods ?? []
            ))
        }
        return .authenticated(try storeSession(from: data))
    }

    /// Finish an MFA login with a TOTP / email / backup code.
    public func verifyMfa(mfaToken: String, code: String) async throws -> AuthSession {
        let body = MfaVerifyRequest(mfaToken: mfaToken, code: code)
        let req = try makeRequest(path: "api/auth/mfa/verify", method: .POST,
                                  body: AnyEncodable(body), tokenMode: true)
        let data = try await send(req)
        return try storeSession(from: data)
    }

    /// Exchange a provider OIDC ID token (Apple / Google) for a session.
    /// A federated provider is the auth factor, so this never returns an MFA
    /// challenge. `name` is only meaningful on Apple's first authorization.
    public func oauthSignIn(provider: String, idToken: String, name: String? = nil) async throws -> AuthSession {
        let body = OAuthSignInRequest(idToken: idToken, name: name)
        let req = try makeRequest(path: "api/auth/oauth/\(provider)", method: .POST,
                                  body: AnyEncodable(body), tokenMode: true)
        let data = try await send(req)
        return try storeSession(from: data)
    }

    /// Request an emailed 6-digit code for the email-MFA path.
    public func sendEmailCode(mfaToken: String) async throws {
        let body = MfaTokenRequest(mfaToken: mfaToken)
        let req = try makeRequest(path: "api/auth/mfa/email/send", method: .POST,
                                  body: AnyEncodable(body))
        try await send(req)
    }

    /// Validate a stored token on launch. Returns the user, or nil when
    /// the server says we're anonymous.
    public func me() async throws -> User? {
        let req = try makeRequest(path: "api/auth/me", method: .GET)
        let data = try await send(req)
        return try decode(MeResponse.self, from: data).user
    }

    /// Re-send the email-verification message for the current (signed-in
    /// but unverified) session. Authed via the Bearer token.
    public func resendVerification() async throws {
        let req = try makeRequest(path: "api/auth/resend-verification", method: .POST)
        try await send(req)
    }

    /// Mark first-run onboarding complete for the current session. Authed
    /// via the Bearer token (CSRF is skipped for token auth server-side).
    public func markOnboarded() async throws {
        let req = try makeRequest(path: "api/account/onboarded", method: .POST)
        try await send(req)
    }

    // ── Plaid (bank linking, Pro-gated) ──────────────────────────────

    /// Current Plaid state: whether the server has credentials, whether the
    /// user is Pro, and their linked items + balances.
    public func plaidStatus() async throws -> PlaidStatus {
        let req = try makeRequest(path: "api/plaid/status", method: .GET)
        return try decode(PlaidStatus.self, from: try await send(req))
    }

    /// Create a Plaid Link token to open the native Link flow. Pass `itemId`
    /// for an update-mode token (re-auth an existing item); set
    /// `accountSelection` to open update mode with account selection (the
    /// NEW_ACCOUNTS_AVAILABLE "add accounts" flow).
    public func plaidLinkToken(itemId: Int? = nil, accountSelection: Bool = false) async throws -> String {
        let req: URLRequest
        if let itemId {
            req = try makeRequest(path: "api/plaid/link/token", method: .POST,
                                  body: AnyEncodable(PlaidLinkTokenBody(itemId: itemId, accountSelection: accountSelection ? true : nil)))
        } else {
            req = try makeRequest(path: "api/plaid/link/token", method: .POST)
        }
        return try decode(PlaidLinkTokenResponse.self, from: try await send(req)).linkToken
    }

    /// After a successful update-mode Link, mark the item repaired (no
    /// public-token exchange happens in update mode).
    public func plaidRepaired(itemId: Int) async throws {
        let req = try makeRequest(path: "api/plaid/item/\(itemId)/repaired", method: .POST)
        _ = try await send(req)
    }

    /// Exchange the Link `public_token` for a stored item (server pulls
    /// balances immediately).
    public func plaidExchange(publicToken: String) async throws {
        let req = try makeRequest(path: "api/plaid/link/exchange", method: .POST,
                                  body: AnyEncodable(PlaidExchangeBody(publicToken: publicToken)))
        _ = try await send(req)
    }

    /// Disconnect a linked item.
    public func plaidRemove(itemId: Int) async throws {
        let req = try makeRequest(path: "api/plaid/item/\(itemId)/remove", method: .POST)
        _ = try await send(req)
    }

    /// Re-pull balances for every linked item; returns the refreshed list.
    public func plaidRefresh() async throws -> [PlaidItem] {
        let req = try makeRequest(path: "api/plaid/refresh", method: .POST)
        return try decode(PlaidItemsResponse.self, from: try await send(req)).items
    }

    /// Revoke the session server-side and clear the local token.
    public func logout() async throws {
        let req = try makeRequest(path: "api/auth/logout", method: .POST)
        // Best-effort: clear locally even if the network call fails.
        defer { tokens.clear() }
        _ = try? await send(req)
    }

    private func storeSession(from data: Data) throws -> AuthSession {
        let r = try decode(SessionResponse.self, from: data)
        guard let token = r.token else {
            throw APIError.decoding("missing token in auth response")
        }
        tokens.set(token)
        return AuthSession(token: token, user: r.user)
    }

    // ── Data sync ─────────────────────────────────────────────────────

    public func fetchData() async throws -> AppData {
        let req = try makeRequest(path: "api/data", method: .GET)
        let data = try await send(req)
        return try decode(AppData.self, from: data)
    }

    public func saveData(_ appData: AppData) async throws {
        let body = DataPutBody(
            bills: appData.bills,
            cards: appData.cards,
            payments: appData.payments,
            accounts: appData.accounts,
            goals: appData.goals,
            transactions: appData.transactions,
            settings: appData.settings
        )
        let req = try makeRequest(path: "api/data", method: .PUT, body: AnyEncodable(body))
        try await send(req)
    }
}
